databuilder/databuilder/models/type_metadata.py

```diff
@@ -0,0 +1,390 @@
+        self.parent = parent
+        self.type_str = type_str
+        self.description = DescriptionMetadata.create_description_metadata(
+            source=None,
+            text=description
+        )
+        self.sort_order: Optional[int] = None
+
+        self._node_iter = self.create_node_iterator()
+        self._relation_iter = self.create_relation_iterator()
+
+    @abc.abstractmethod
+    def __eq__(self, other: Any) -> bool:
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def is_terminal_type(self) -> bool:
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def create_node_iterator(self) -> Iterator[GraphNode]:
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def create_relation_iterator(self) -> Iterator[GraphRelationship]:
+        raise NotImplementedError
+
+    def create_next_node(self) -> Optional[GraphNode]:
+        try:
+            return next(self._node_iter)
+        except StopIteration:
+            return None
+
+    def create_next_relation(self) -> Optional[GraphRelationship]:
+        try:
+            return next(self._relation_iter)
+        except StopIteration:
+            return None
+
+    def key(self) -> str:
+        if isinstance(self.parent, ColumnMetadata):
+            return f"{self.parent_key()}/type/{self.name}"
+        else:
+            return f"{self.parent_key()}/{self.name}"
+
+    def description_key(self) -> str:
+        if self.description:
+            description_id = self.description.get_description_id()
+            return f"{self.key()}/{description_id}"
+        else:
+            return ''
+
+    def parent_key(self) -> str:
+        if isinstance(self.parent, ColumnMetadata):
+            column_key = self.parent.get_column_key()
+            return column_key if column_key else ''
+        else:
+            return self.parent.key()
+
+    def parent_label(self) -> str:
+        if isinstance(self.parent, ColumnMetadata):
+            return ColumnMetadata.COLUMN_NODE_LABEL
+        else:
+            return TypeMetadata.NODE_LABEL
+
+    def __repr__(self) -> str:
+        return f"TypeMetadata({self.type_str!r})"
+
+
+class ArrayTypeMetadata(TypeMetadata):
+    kind = 'array'
+
+    def __init__(self, *args: Any, **kwargs: Any) -> None:
+        super(ArrayTypeMetadata, self).__init__(*args, **kwargs)
+        self.data_type: Optional[TypeMetadata] = None
+
+    def __eq__(self, other: Any) -> bool:
+        if isinstance(other, ArrayTypeMetadata):
+            return (self.name == other.name and
+                    self.type_str == other.type_str and
+                    self.description == other.description and
+                    self.sort_order == other.sort_order and
+                    self.data_type == other.data_type and
+                    self.key() == other.key())
+        return False
+
+    def is_terminal_type(self) -> bool:
+        return isinstance(self.data_type, ScalarTypeMetadata)
```

Comment:
put a comment here, maybe

Reply:
oh I also realized a thing. Maybe we should let people specify is_terminal_type from the constructor.